src/main/java/org/ibp/api/java/impl/middleware/dataset/StudyBookTableBuilder.java

```diff
@@ -85,4 +86,65 @@ Integer getDuplicatedFoundNumber() {
 		return this.duplicatedFoundNumber;
 	}
 
+	public Table<String, String, String> buildEnvironmentVariablesTable(final List<List<String>> data, final List<MeasurementVariable> datasetMeasurementVariables) throws ApiRequestValidationException {
+
+		this.duplicatedFoundNumber = 0;
+
+		final BindingResult
+				errors = new MapBindingResult(new HashMap<String, String>(), ObservationsPutRequestInput.class.getName());
+
+		final List<String> headers = data.get(0);
+		final List<List<String>> values = data.subList(1, data.size());
+
+		// check that headers contains TRIAL_INSTANCE
+		if (!headers.contains(TRIAL_INSTANCE)) {
+			errors.reject("required.header.trial.instance", null , "");
+			throw new ApiRequestValidationException (errors.getAllErrors());
+		}
+
+		// filter measurement variables from header
+		final List<Integer> importMeasurementVariablesIndex = new ArrayList<>();
+
+		for (final String header: headers) {
```

Comment:
Improve this code to avoid code duplicate. You can extract this to a method

```
List<Integer> importMeasurementVariablesIndex = this.filterImportMeasurementVariables(headers, datasetMeasurementVariables, errors);

 private List<Integer> filterImportMeasurementVariables(
            List<String> headers,
            List<MeasurementVariable> datasetMeasurementVariables,
            BindingResult errors
    ) {
        List<Integer> importMeasurementVariablesIndex = new ArrayList<>();
        for (String header : headers) {
            for (MeasurementVariable measurementVariableDto : datasetMeasurementVariables) {
                if (measurementVariableDto.getName().equals(header) || measurementVariableDto.getAlias().equals(header)) {
                    int headerIndex = headers.indexOf(header);
                    if (importMeasurementVariablesIndex.contains(headerIndex)) {
                        errors.reject("duplicated.measurement.variables.not.allowed", null, "");
                        throw new ApiRequestValidationException(errors.getAllErrors());
                    } else {
                        importMeasurementVariablesIndex.add(headerIndex);
                        break;
                    }
                }
            }
        }
        return importMeasurementVariablesIndex;
    }
```